scripts/sensor/03b-time_frequency_csp.py

```diff
@@ -0,0 +1,866 @@
+
+        msg = (f"We will split the epoch time interval"
+               f" into {n_time_windows} windows.")
+        logger.info(gen_log_message(msg, step=3))
+        times = np.linspace(
+            cfg.epochs_tmin, cfg.epochs_tmax, n_time_windows + 1)
+        time_ranges = list(zip(times[:-1], times[1:]))
+        centered_w_times = (times[1:] + times[:-1])/2
+
+        # Recap:
+        assert n_time_windows == n_times - 1
+        assert len(time_ranges) == n_time_windows
+        assert n_freq_windows == cfg.n_freqs - 1
+        assert len(freq_ranges) == n_freq_windows
+
+        self.freqs = freqs
+        self.freq_ranges = freq_ranges
+        self.times = times
+        self.time_ranges = time_ranges
+        self.centered_w_times = centered_w_times
+        self.n_time_windows = n_time_windows
+        self.n_freq_windows = n_freq_windows
+
+        # TODO: Check that each subject contains the 2 conditions.
+        # in order to avoid bad news after 1 hour of computation ?
+
+
+def prepare_labels(*, epochs: BaseEpochs, cfg) -> np.ndarray:
+    """Return the projection of the events_id on a boolean vector.
+
+    This projection is useful in the case of hierarchical events:
+    we project the different events contain in one condition into
+    just one label.
+
+    Returns:
+    --------
+    A boolean numpy array containing the labels.
+    """
+    tf_conditions = cfg.time_frequency_conditions
+    assert len(tf_conditions) == 2
+
+    # TODO: this creates sometime a critical error sometime
+    # but difficult to debug
+    # because there is not the subject number written in the error.
+    # "A critical error occurred. The error message was:
+    # 'Event "3_item" is not in Epochs.
+    # Event_ids must be one of "1_item/medium,
+    # 1_item/short" The epochs.metadata Pandas query did not yield any
+    # results: invalid syntax'"
+    epochs_cond_0 = epochs[tf_conditions[0]]
+    event_id_condition_0 = set(epochs_cond_0.events[:, 2])
+    epochs_cond_1 = epochs[tf_conditions[1]]
+    event_id_condition_1 = set(epochs_cond_1.events[:, 2])
+
+    y = epochs.events[:, 2].copy()
+    for i in range(len(y)):
+        if y[i] in event_id_condition_0:
+            y[i] = 0
+        elif y[i] in event_id_condition_1:
+            y[i] = 1
+        else:
+            msg = (f"Event_id {y[i]} is not contained in "
+                   f"{tf_conditions[0]}'s set {event_id_condition_0}  nor in "
+                   f"{tf_conditions[1]}'s set {event_id_condition_1}.")
+            # TODO: Add subject number in this error?
+            logger.warning(msg)
+    return y
+
+
+def prepare_epochs_and_y(
+    *,
+    epochs: BaseEpochs,
+    cfg,
+    fmin: float,
+    fmax: float
+) -> Tuple[BaseEpochs, np.ndarray]:
+    """Band-pass and clean the epochs and prepare labels.
+
+    Returns:
+    --------
+    epochs_filter, y
+    """
+    # Prepare epoch_filter
+    epochs_filter = epochs.copy()
+    epochs_filter.decimate(decim=cfg.decim)
```

Comment:
decim should already have been applied to clean epochs I think

Reply:
Yes exactly, the decim parameter was already applied, but it's possible that to speed up the calculation we want to decim even further. But we should do this without the cfg.decim parameter and do this automatically.

Or... Do you think that by not decimating, we could facilitate the estimation of the covariance ?

Reply:
don't do this and profile the code to see where time is spent.

Reply:
![image](https://user-images.githubusercontent.com/38733898/124887860-dc2e9300-dfd5-11eb-91a6-606ca277fcfd.png)

So the majority of the is spent in the CSP in the covariance estimator.